src/algebra/module/zlattice.lean

```diff
@@ -0,0 +1,229 @@
+/-
+Copyright (c) 2023 Xavier Roblot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Xavier Roblot
+-/
+import measure_theory.group.fundamental_domain
+
+/-!
+# ℤ-lattices
+
+Let `E` be a finite dimensional vector space over a `normed_linear_ordered_field` `K` with a solid
+norm and that is also a `floor_ring`, e.g. `ℚ` or `ℝ`. A (full) ℤ-lattice `L` of `E` is a discrete
+subgroup of `E` such that `L` spans `E` over `K`.
+
+The ℤ-lattice `L` can be defined in two ways:
+* For `b` a basis of `E`, then `submodule.span ℤ (set.range b)` is a ℤ-lattice of `E`.
+* As an `add_subgroup E` with the additional properties:
+  `∀ r : ℝ, (L ∩ metric.closed_ball 0 r).finite`, that is `L` is discrete
+  `submodule.span ℝ (L : set E) = ⊤`, that is `L` spans `E` over `K`.
+
+## Main result
+* `zspan.is_add_fundamental_domain`: for a ℤ-lattice `submodule.span ℤ (set.range b)`, proves that
+the set defined by `zspan.fundamental_domain` is a fundamental domain.
+
+-/
+
+open_locale big_operators
+
+noncomputable theory
+
+namespace zspan
+
+open measure_theory measurable_set submodule
+
+variables {E ι : Type*}
+
+section normed_lattice_field
+
+variables {K : Type*} [normed_linear_ordered_field K]
+variables [normed_add_comm_group E] [normed_space K E]
+variables (b : basis ι K E)
+
+/-- The fundamental domain of the ℤ-lattice spanned by `b`. See `zspan.is_add_fundamental_domain`
+for the proof that it is the fundamental domain. -/
+def fundamental_domain : set E := { m | ∀ i, b.repr m i ∈ set.Ico (0 : K) 1 }
+
+@[simp]
+lemma mem_fundamental_domain {m : E} :
+  m ∈ fundamental_domain b ↔ ∀ i, b.repr m i ∈ set.Ico (0 : K) 1 := iff.rfl
+
+variables [floor_ring K]
+
+section fintype
+
+variable [fintype ι]
+
+/-- The map that sends a vector of `E` to the element of the ℤ-lattice spanned by `b` obtained
+by rounding down its coordinates on the basis `b`. -/
+def floor (m : E) : span ℤ (set.range b) := ∑ i, ⌊b.repr m i⌋ • b.restrict_scalars ℤ i
+
+/-- The map that sends a vector of `E` to the element of the ℤ-lattice spanned by `b` obtained
+by rounding up its coordinates on the basis `b`. -/
+def ceil (m : E) : span ℤ (set.range b) := ∑ i, ⌈b.repr m i⌉ • b.restrict_scalars ℤ i
```

Comment:
Can you add the lemma that `(m : span ℤ (set.range b)) : ceil m = m`? That is, that `ceil` (and `floor`) is surjective?

Reply:
I added two new lemma: 
```lean
lemma floor_eq_self_of_mem (m : E) (h : m ∈ span ℤ (set.range b)) : (floor b m : E) = m  
lemma ceil_eq_self_of_mem (m : E) (h : m ∈ span ℤ (set.range b)) : (ceil b m : E) = m
```
It was convenient for the proofs to add two news lemmas to `algebra.order.floor`:
```lean
@[simp]
lemma int.floor_eq_self (h : ∃ z : ℤ, (z : k) = b) : (⌊b⌋ : k) = b :=
by { obtain ⟨z, rfl⟩ := h; exact galois_connection.l_u_l_eq_l int.gc_coe_floor z, }

@[simp]
lemma int.ceil_eq_self (h : ∃ z : ℤ, (z : k) = b) : (⌈b⌉ : k) = b :=
by { obtain ⟨z, rfl⟩ := h; exact galois_connection.u_l_u_eq_u int.gc_ceil_coe z, }
```